certora/specs/Account.spec

```diff
@@ -0,0 +1,182 @@
+
+    f(e, args);
+
+    assert call => (
+        (
+            f.selector == sig:execute(bytes32,bytes).selector &&
+            isEntryPointOrSelf
+        ) || (
+            f.selector == sig:executeFromExecutor(bytes32,bytes).selector &&
+            isExecutionModule
+        ) || (
+            // Note: rule callInstallModule checks that the call target is the module being installed
+            f.selector == sig:installModule(uint256,address,bytes).selector &&
+            isEntryPointOrSelf &&
+            call_value == 0
+        ) || (
+            // Note: rule callUninstallModule checks that the call target is the module being uninstalled
+            f.selector == sig:uninstallModule(uint256,address,bytes).selector &&
+            isEntryPointOrSelf &&
+            call_value == 0
+        ) || (
+            f.selector == sig:validateUserOp(Account.PackedUserOperation,bytes32,uint256).selector &&
+            isEntryPoint &&
+            (
+                (
+                    // payPrefund (target is entryPoint and argsLength is 0)
+                    call_target == entryPoint() &&
+                    call_value > 0 &&
+                    call_argsLength == 0
+                ) || (
+                    // isValidSignatureWithSender (target is as validation module)
+                    isModuleInstalled(1, call_target, context) &&
+                    call_value == 0
+                )
+            )
+        ) || (
+            f.isFallback &&
+            call_target == getFallbackHandler(to_bytes4(call_selector)) &&
+            call_value == e.msg.value
+        )
+    );
+}
+
+rule callInstallModule(env e, uint256 moduleTypeId, address module, bytes initData) {
+    installModule(e, moduleTypeId, module, initData);
+
+    assert call        == true;
+    assert call_target == module;
+    assert call_value  == 0;
+}
+
+rule callUninstallModule(env e, uint256 moduleTypeId, address module, bytes deInitData) {
+    uninstallModule(e, moduleTypeId, module, deInitData);
+
+    assert call        == true;
+    assert call_target == module;
+    assert call_value  == 0;
+}
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│                                                 DELEGATECALL OPCODE                                                 │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+ghost bool    delegatecall;
+ghost address delegatecall_target;
+ghost uint256 delegatecall_argsLength;
+
+hook DELEGATECALL(uint256 gas, address target, uint256 argsOffset, uint256 argsLength, uint256 retOffset, uint256 retLength) uint256 rc {
+    if (executingContract == currentContract) {
+        delegatecall = true;
+        delegatecall_target = target;
+        delegatecall_argsLength = argsLength;
+    }
+}
+
+rule delegatecallOpcodeRule(
```

Comment:
This rule doesn't work. Some function can trigger CALL that HAVOC the ghosts ...